paimon-core/src/main/java/org/apache/paimon/index/SimpleHashBucketAssigner.java

```diff
@@ -83,7 +88,15 @@ public int assign(int hash) {
 
             Long num = bucketInformation.computeIfAbsent(currentBucket, i -> 0L);
             if (num >= targetBucketRowNumber) {
-                loadNewBucket();
+                if (-1 != maxBucketsNum && bucketInformation.size() >= maxBucketsNum) {
+                    int bucket =
+                            KeyAndBucketExtractor.bucketWithUpperBound(
+                                    bucketInformation.keySet(), hash, maxBucketsNum);
+                    hash2Bucket.put(hash, (short) bucket);
```

Comment:
Add a util function of PartitionIndex#cacheBucketAndGet, then here use?

Reply:
done :)